src/main/java/reactor/netty/resources/ConnectionProvider.java

```diff
@@ -372,10 +373,33 @@ public final SPEC maxLifeTime(Duration maxLifeTime) {
 		 * </pre>
 		 * <p>By default this is not enabled.
 		 *
+		 * If enabled, this method is providing a {@link DefaultPooledConnectionProviderMeterRegistrar} to define which meters are going to be collected.
+		 *
 		 * @param metricsEnabled true enables metrics collection; false disables it
 		 * @return {@literal this}
 		 */
 		public final SPEC metrics(boolean metricsEnabled) {
+			return this.metrics(metricsEnabled, new DefaultPooledConnectionProviderMeterRegistrar());
+		}
+
+		/**
+		 * Whether to enable metrics to be collected and registered in Micrometer's
+		 * {@link io.micrometer.core.instrument.Metrics#globalRegistry globalRegistry}
+		 * under the name {@link reactor.netty.Metrics#CONNECTION_PROVIDER_PREFIX}. Applications can
+		 * separately register their own
+		 * {@link io.micrometer.core.instrument.config.MeterFilter filters} associated with this name.
+		 * For example, to put an upper bound on the number of tags produced:
+		 * <pre class="code">
+		 * MeterFilter filter = ... ;
+		 * Metrics.globalRegistry.config().meterFilter(MeterFilter.maximumAllowableTags(CONNECTION_PROVIDER_PREFIX, 100, filter));
+		 * </pre>
+		 * <p>By default this is not enabled.
+		 *
+		 * @param metricsEnabled true enables metrics collection; false disables it
+		 * @param registrar the registrar used to define what meters are going to be collected.
+		 * @return {@literal this}
+		 */
+		public final SPEC metrics(boolean metricsEnabled, PooledConnectionProvider.MeterRegistrar registrar) {
```

Comment:
> The idea of the new method is to remove the dependency on Micrometer.

see my comment above

> The user will provide an implementation of the second parameter (it is better to be a Supplier<> so that we will call it only when the metrics are enabled)

if the user calls this with `false`, he/she is more likely to use the overload with only one parameter anyway, so we don't understand what using a Supplier gives in terms of benefits. Registrars are meant to be lightweight anyway. It's actually calling the registration method that is costly, not creating the registrar, no?

Reply:
In some environments metrics can be activated (test/canary) but on the production disabled. The configuration can be controlled by an environment variable or something else (system property, configuration service). To avoid many if/else clauses we offer boolean parameter to enable/disable the whole feature and when it is disabled the instance will not be created. Similar features were requested by `Spring Cloud Gateway` for example `HttpClient.wiretap`, `HttpClient.keepAlive` etc.
With Supplier we the instance will be created only when it is needed.

```
boolean metricsDisabled = Boolean.parseBoolean(System.getProperty("metricsEnabled", "false"));

ConnectionProvider.builder("test")
                  .maxConnections(2)
                  .metrics(metricsDisabled, () -> ...)
                  .forRemoteHost(InetSocketAddress.createUnresolved("localhost", server.port()),
                                 spec -> spec.maxConnections(1).metrics(metricsDisabled, () -> ...))
                  .build();
```

Reply:
Gotcha. Still, the registrar should be lightweight. Using a Supplier seems to go against the argument you made in https://github.com/reactor/reactor-netty/pull/1218#discussion_r459312327

Reply:
Do you mean that the instance will be created even without calling the `Supplier.get()`?

Reply:
> Registrars are meant to be lightweight anyway. 

We do not control the user's implementation, we do not know whether it is lightweight or not.